### PR TITLE
performace improvement for moveByReference, rebuild, create and delete nested tree records

### PR DESCRIPTION
**Pull request by shandak**
### Summary of Changes
Every nested tree node movement or rebuild usually executes quite a long time. Especially for the supertable `#__assets`. 

New `moveByReference` changes will increase performance by changing only leaves which between a new and an old node positions and does it only in one query(only one update query). We don't need to compress the tree several times anymore. New changes only work with data which actually required for changes.

I got an enormous time execution difference inside my project with this changes.
Before it was executed by 587 seconds 
and the same action with my changes was executed only by 18 seconds.

The query has a format like:

 ```UPDATE `#__reditem_categories`
SET lft = CASE WHEN lft BETWEEN 2 AND 3 THEN lft+4 WHEN lft BETWEEN 2 AND 7 THEN lft-2 ELSE lft END
	, rgt = CASE WHEN rgt BETWEEN 2 AND 3 THEN rgt+4 WHEN rgt BETWEEN 2 AND 7 THEN rgt-2 ELSE rgt END
WHERE (lft BETWEEN 2 AND 7 OR rgt BETWEEN 2 AND 7)```

New `rebuild` changes are also improving the process dramatically. It was tested with a broken `#__asset` table with more than 1 million records. I got the next results:
For full broken tree
before changes 671 sec
after my changes 39 sec

New `delete` and `create` changes just improve a process a bit. I've changed them for completely avoid an old approach  related with `_getTreeRepositionData`. I completely removed it.

### Testing Instructions
Every movement, rebuild, create and delete of nested tree entities(like category) must be as usual.

### Expected result
Performance improvement for huge nested tree tables must be substantially fast than right now.

### Actual result
Every nested tree node movement, rebuild usually executes quite a long time.

### CODE REVIEW

File: libraries/src/Table/Nested.php [hunk 1]
```diff
@@ -424,52 +383,65 @@ public function moveByReference($referenceId, $position = 'after', $pk = null, $
...
+		}
 
-		// Move the nodes back into position in the tree using the calculated offsets.
 		$query->clear()
-			->update($this->_tbl)
-			->set('rgt = ' . (int) $offset . ' - rgt')
-			->set('lft = ' . (int) $offset . ' - lft')
-			->set('level = level + ' . (int) $levelOffset)
-			->where('lft < 0');
+			->update($this->_db->qn($this->_tbl))
```
Comment (csthomas) [inline libraries/src/Table/Nested.php]: Change the name of the `qn` method to 'quoteName`.

File: libraries/src/Table/Nested.php [hunk 2]
```diff
@@ -424,52 +383,87 @@ public function moveByReference($referenceId, $position = 'after', $pk = null, $
...
-		$query->clear()
-			->update($this->_tbl)
-			->set('lft = lft + ' . (int) $node->width)
-			->where($repositionData->left_where);
-		$this->_db->setQuery($query);
+		// The width of node that is being moved
+		$width = $node->width;
 
-		$this->_runQuery($query, 'JLIB_DATABASE_ERROR_MOVE_FAILED');
+		// The distance that our node will travel to reach it's destination
```
Comment (Quy) [inline libraries/src/Table/Nested.php]: Change to `its`

**Comment by Hackwar** [comment]
Thank you for working on the nested sets code. I fear that this will not be really backwards compatible with the changed method names and thus would aim this directly at the 4.0 branch to be safe there. 

While you are at it, could you also look into the code to store new ordering of items? Last time I looked, that code did some nasty stuff by simply deleting all rgt and lft values for the items to order, setting the order column to the values and then executing a rebuild(), instead of executing the right nested sets queries directly.

**Comment by shandak** [comment]
Thank you @Hackwar for your's feedback.
With regard to back compatibility, you told about removed `_getTreeRepositionData` yeah? If so I can turn it back and mark as deprecated. Will it be enough to release the code before 4.0? It seems it must be fully compatible with the current one.
And yeah, I'll rebuild 'new ordering of items' in the next PR.

**Comment by shandak** [comment]
@Hackwar I turned back `_getTreeRepositionData` already

**Comment by csthomas** [comment]
This can be very good improvement, I will try to find a time to test it.

File: libraries/src/Table/Nested.php [hunk 3]
```diff
@@ -1294,71 +1300,182 @@ public function rebuild($parentId = null, $leftId = 0, $level = 0, $path = '')
...
+	 * Method to recursively rebuild the whole nested set tree.
+	 *
+	 * @param   object        $parent  The root of the tree to rebuild.
+	 * @param   integer       $lft     The left id to start with in building the tree.
+	 * @param   integer       $level   The level to assign to the current nodes.
+	 * @param   string|false  $path    The path to the current nodes or false if table has no 'path' field.
+	 *
+	 * @return  integer  1 + value of root rgt on success, false on failure
+	 *
+	 * @since   11.1
```
Comment (Quy) [inline libraries/src/Table/Nested.php]: Change to `__DEPLOY_VERSION__`

**Comment by richard67** [comment]
@shandak Could you update your branch with current staging and solve conflicts in file `libraries/src/Table/Nested.php`? Thanks in advance.

**Comment by shandak** [comment]
@richard67 conflicts solved

File: tests/unit/suites/libraries/joomla/table/JTableNestedTest.php [hunk 4]
```diff
@@ -658,6 +658,64 @@ public function test_getNode()
 		$this->assertFalse($node);
 	}
 
+	/**
+	 * Tests the `_getTreeRepositionValues` method.
+	 *
+	 * @return  void
+	 *
+	 * @since   12.1
```
Comment (PhilETaylor) [inline tests/unit/suites/libraries/joomla/table/JTableNestedTest.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

**Comment by laoneo** [comment]
Sorry that it toke so long to respond. Performance improvements are always welcome in Joomla. But it needs proper testing, especially here as the changes do affect many extensions. As @Hackwar already mentioned, can you please rebase the pr on the branch 4.2-dev? In the meantime I'm closing the pr. When ready please reopen so we can get it tested. Thanks for your help making Joomla better.

**Pull request closed without merging**